### PR TITLE
[transformers_modeling_backend] Populate rotary inv_freq in _init_weights

### LINKED ISSUE #4783: [transformers_modeling_backend]Regression of #3775: rotary inv_freq is left uninitialized after `to_empty()`
#### Bug description

This is a regression of the RoPE part of #3775. HF rotary modules compute `inv_freq` in `__init__`, so with the backend's `meta` init + `to_empty()` the buffer is materialized as uninitialized memory, and nothing puts real values back. RoPE then rotates by whatever was in that memory.

It fails silently. Nothing raises, `inv_freq` is non-persistent so a checkpoint load does not restore it either, and the only symptom is loss starting at near-random -- the 3.63 vs 11.08 gap measured in #3775. 

#### Root cause: the Transformers pin moved to 5.x

#3772 fixed this with a [compatibility loop](https://github.com/pytorch/torchtitan/blob/183efed45d7d8d2cd880dd3803a3334ed47aba60/torchtitan/experiments/transformers_modeling_backend/model.py#L1372-L1380) in `init_states()` that recomputes `inv_freq` from `module.rope_init_fn`. That worked against the Transformers 4.57.1 pin CI used at the time. #2679 then moved the pin to 5.9.0 and #4404 to the current 5.10.1, and Transformers 5.0.0 demoted `rope_init_fn` from an instance attribute to a local variable inside `__init__`. `getattr(..., None)` returns `None`, so the loop has been a no-op ever since.

Transformers 5.x would otherwise cover this in the `RotaryEmbedding` branch of `PreTrainedModel._init_weights`, but the backend
[replaces that method](https://github.com/pytorch/torchtitan/blob/183efed45d7d8d2cd880dd3803a3334ed47aba60/torchtitan/experiments/transformers_modeling_backend/model.py#L1044-L1047) with `_init_weights_patched`, which has no rotary branch. So, neither path runs.

#### Repro

CPU-only, no download and no checkpoint. The rotary buffers are filled with NaN after `to_empty()` so that "not reinitialized" is deterministic rather than dependent on the contents of uninitialized memory.

```python
import torch
from transformers import LlamaConfig
from transformers.models.llama import modeling_llama
from transformers.models.llama.modeling_llama import (
    LlamaForCausalLM,
    LlamaRotaryEmbedding,
)
from torchtitan.experiments.transformers_modeling_backend.model import (
    HFTransformerModel,
)

config = LlamaConfig(
    hidden_size=64,
    intermediate_size=128,
    num_attention_heads=4,
    num_key_value_heads=4,
    num_hidden_layers=2,
    vocab_size=128,
    max_position_embeddings=256,
)

# Ground truth: the same module built straight on CPU, where HF's __init__
# computes inv_freq normally.
expected = LlamaRotaryEmbedding(config).inv_freq.clone()

HFTransformerModel._patch_hf_llama_like(
    None,
    decoder_layer_cls=modeling_llama.LlamaDecoderLayer,
    attention_cls=modeling_llama.LlamaAttention,
    mlp_cls=modeling_llama.LlamaMLP,
    experts_cls=None,
    router_cls=None,
)

with torch.device("meta"):
    model = LlamaForCausalLM(config)
model.to_empty(device="cpu")

rotary = model.model.rotary_emb
rotary.inv_freq.fill_(float("nan"))
rotary.original_inv_freq.fill_(float("nan"))

model.apply(model._init_weights)

print("rope_init_fn attribute exists:", hasattr(rotary, "rope_init_fn"))
print("inv_freq still NaN:", bool(torch.isnan(rotary.inv_freq).all()))
print("inv_freq matches expected:", bool(torch.allclose(rotary.inv_freq, expected)))
```

Output:

```text
rope_init_fn attribute exists: False
inv_freq still NaN: True
inv_freq matches expected: False
```

#### Proposed fix

Reproduce the Transformers rotary initialization inside `_init_weights_patched` instead of patching around it afterwards, so `init_states()` covers rotary modules like any other submodule and there is only one place to keep in sync with Transformers.

#### Proposed regression test

A unit test over `meta` -> `to_empty()` -> `init_states()` that checks `inv_freq` against the same rotary module instantiated normally on CPU, where HF's `__init__` computes the buffer itself. Taking the expected values from that path keeps them independent of the initialization under test. The point is that the next Transformers bump should fail CI instead of silently disabling RoPE again -- that is exactly how this regressed.

I have a patch along those lines, with regression tests covering default and llama3 rope. Happy to open a PR if that is useful.

cc @HosseinKaviani-H 

### Versions

- TorchTitan `main` at `183efed45d7d`
- Transformers 5.10.1 (the current backend CI pin); applies to every 5.x, since
  `self.rope_init_fn` has been gone since 5.0.0
- PyTorch `2.14.0a0+git6ebda8c`
- CPU

### PR DESCRIPTION
**Pull request by Jiarwang77**
Fixes #4783.

### Issue description

`HFTransformerModel.init_states()` did not write real values into `inv_freq` on HF rotary modules after `meta` init + `to_empty()`, so RoPE rotated by uninitialized memory. The compatibility loop added in #3772 reads `module.rope_init_fn`, which Transformers 5.0.0 changed from an instance attribute to a local variable inside `__init__`, so the loop has been a no-op since #2679 moved the pin to 5.x. Transformers' own `RotaryEmbedding` branch in `PreTrainedModel._init_weights` does not run either, because the backend replaces all of that method.

Nothing raises, and `inv_freq` is non-persistent, so loading a checkpoint does not write it either.

### Fix

- **`model.py`** -- reproduce the Transformers rotary initialization inside `_init_weights_patched`, so `init_states()` reaches it like any other submodule, and drop the loop that no longer does anything. The guard is `original_inv_freq`, the same one Transformers uses, so the layer-typed rotary modules handled by model-specific `_init_weights` overrides are also skipped here. Both buffers are written, because `dynamic` and `longrope` re-register `inv_freq` from `original_inv_freq`.
- **`tests/test_rope_init.py`** (new) -- CPU unit tests over `meta` -> `to_empty()` -> `init_states()`. They compare against a rotary module built directly on CPU, so the expected values come from HF's own `__init__` rather than from the code under test.
- **`.github/workflows/unit_test_cpu_transformers_modeling_backend.yaml`** (new) -- `unit_test_cpu.yaml` has `paths-ignore: torchtitan/experiments/**` and runs `pytest tests/unit_tests/cpu`, and the 8-GPU workflow only runs `integration_tests.py`, so without this nothing would run the tests. Modelled on `unit_test_cpu_torchft.yaml`. It installs from `.ci/docker/requirements-transformers-modeling-backend.txt` and lists that file under `paths`, so a Transformers bump triggers this job and runs the tests against the new version -- which is how this regressed in the first place.

### CODE REVIEW

**Comment by tianyu-l** [comment]
@claude review this PR. Evaluate if this is the most proper fix.

**Comment by tianyu-l** [comment]
@Jiarwang77 do you agree with review claude gave? Seems that there could be some improvements.

@HosseinKaviani-H could you also help take a look?

File: .github/workflows/unit_test_cpu_transformers_modeling_backend.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        # torchvision must come from the same nightly channel as torch, because
+        # transformers depends on it transitively via image_utils.
+        pip install --force-reinstall --pre torch torchvision --index-url https://download.pytorch.org/whl/nightly/cpu
+
+        USE_CPP=0 python -m pip install --pre torchao --index-url https://download.pytorch.org/whl/nightly/cpu
+
+        # Install from the pin file rather than a literal version, so that a
+        # Transformers bump has to run these tests against the new version.
+        python -m pip install -r .ci/docker/requirements-transformers-modeling-backend.txt
```
Comment (HosseinKaviani-H) [inline .github/workflows/unit_test_cpu_transformers_modeling_backend.yaml]: Pin-file install is right, but integration_test_8gpu_transformers_modeling_backend.yaml:87 still hardcodes transformers==5.9.0 vs 5.10.1 here. Two versions tested, you should unify.

File: torchtitan/experiments/transformers_modeling_backend/model.py [hunk 2]
```diff
@@ -1043,6 +1044,30 @@ def _init_weights_patched(self, module):
...
+                # overrides are also skipped here.
+                # ROPE_INIT_FUNCTIONS has no "default" entry; the default is a
+                # method on the rotary class instead.
+                rope_init_fn = (
+                    module.compute_default_rope_parameters
+                    if module.rope_type == "default"
+                    else ROPE_INIT_FUNCTIONS[module.rope_type]
+                )
+                inv_freq, _ = rope_init_fn(module.config, module.inv_freq.device)
+                module.inv_freq.copy_(inv_freq)
```
Comment (HosseinKaviani-H) [inline torchtitan/experiments/transformers_modeling_backend/model.py]: Under TP, is module.inv_freq a DTensor here? parallelize() runs before to_empty()/init_weights() and hf_sharding.py:437 declares these buffers replicated. Copying a plain tensor into a DTensor would raise an error then cc @tianyu-l

File: torchtitan/experiments/transformers_modeling_backend/model.py [hunk 3]
```diff
@@ -1043,6 +1044,30 @@ def _init_weights_patched(self, module):
...
+                # rotary modules handled by model-specific _init_weights
+                # overrides are also skipped here.
+                # ROPE_INIT_FUNCTIONS has no "default" entry; the default is a
+                # method on the rotary class instead.
+                rope_init_fn = (
+                    module.compute_default_rope_parameters
+                    if module.rope_type == "default"
+                    else ROPE_INIT_FUNCTIONS[module.rope_type]
+                )
+                inv_freq, _ = rope_init_fn(module.config, module.inv_freq.device)
```
Comment (HosseinKaviani-H) [inline torchtitan/experiments/transformers_modeling_backend/model.py]: Your comment says this branch is reproduced from Transformers, but HF calls rope_fn(module.config) with no device arg and uses init.copy_ rather than Tensor.copy_. Is this intentional?
